pkg/plugins/runtime/k8s/plugin.go

```diff
@@ -128,11 +134,45 @@ func addPodReconciler(mgr kube_ctrl.Manager, rt core_runtime.Runtime, converter
 			Zone:              rt.Config().Multizone.Remote.Zone,
 			ResourceConverter: converter,
 		},
+		Persistence:     dns.NewMeshedPersistence(rt.ConfigManager()),
 		SystemNamespace: rt.Config().Store.Kubernetes.SystemNamespace,
 	}
 	return reconciler.SetupWithManager(mgr)
 }
 
+func addDNS(mgr kube_ctrl.Manager, rt core_runtime.Runtime) error {
+	ipam, err := dns.NewSimpleIPAM(rt.Config().DNSServer.CIDR)
+	if err != nil {
+		return err
+	}
+	p := dns.NewMeshedPersistence(rt.ConfigManager())
+	vipsSync, err := dns.NewVIPsSynchronizer(rt.DNSResolver(), p, rt.LeaderInfo())
+	if err != nil {
+		return err
+	}
+	server, err := dns.NewDNSServer(rt.Config().DNSServer.Port, rt.DNSResolver(), rt.Metrics())
+	if err != nil {
+		return err
+	}
+	if err := rt.Add(server, vipsSync); err != nil {
+		return err
+	}
```

Comment:
Ok, I see why this is a case. I think it's not a good idea to carry all the internals of DNS to the reconciler. What you really want to do is to Instruct `VipAllocator` to do the update on the mesh X. So my proposal is
1) Include VipAllocator into Runtime
2) Instead of having
```
func CreateOrUpdateVIPConfig(p *MeshedPersistence, rm manager.ReadOnlyResourceManager, r DNSResolver, ipam IPAM, mesh string) error {
```
This should be
```
func (d *vipsAllocator) CreateOrUpdateVIPConfig(mesh string) error {
```
3) Pass VIP allocator to this reconciller and execute this method
4) Move back all DNS component preparation into common components

Reply:
I didn't want to put VipAllocator into Runtime. I made it stateless and dependent only on runtime components, so now it's easy to create it from any place:
```go
vipsAllocator, err := dns.NewVIPsAllocator(
	rt.ReadOnlyResourceManager(),
	rt.ConfigManager(),
	rt.Config().DNSServer.CIDR,
	rt.DNSResolver(),
)
```